### PR TITLE
[RL, Spec] Introduce top-p mask capture for spec and add DFlash/DSpark impl

### PR DESCRIPTION
**Pull request by jvmncs**
## Relationship to existing work

#27408 introduced sampling-mask output support. #35205 restored the finite-`top_k` requirement and bounded support packing for sampling masks. #33459 added the DFlash SpecV2 logprob path.

This PR extends that machinery to the DFlash family, with worker-specific adapters for DFlash and DSpark, while preserving the finite-`top_k` contract from #35205.

## Motivation

Sampling masks let RL trainers reconstruct the inference policy more accurately and reduce rollout/training policy mismatch. This change makes that metadata available when using DFlash or DSpark speculative decoding, and introduces a contract that other spec implementations can implement for their own support.

## Modifications

- Add a shared speculative sampling-mask capture/output contract used by DFlash and DSpark.
- Capture masks and selected-token logprobs from the final adjusted target policy, including grammar, penalties, logit bias, and custom logit processors.
- Keep speculative acceptance logic method-specific.
- Align metadata with committed `out_tokens` and `commit_lens`, letting the existing result-processing path trim it alongside generated tokens.
- Support DSpark compact/ragged verification after logits are restored to strided layout.
- Reconstruct target policy per row for mixed-greedy DSpark batches; greedy rows emit singleton masks with logprob 0.
- Pack only finite top-k support before device-to-host transfer.
- Reject sampling-mask requests with unbounded top-k, `min_p`, simulated acceptance, or unsupported speculative methods.
- For DFlash, additionally require non-greedy verification support and exact acceptance thresholds, and prevent incompatible runtime threshold changes while mask requests are active.
- Do not enable sampling-mask capture for EAGLE, NGRAM, or other speculative methods.

## Tests

 - Added DFlash end-to-end coverage in the existing DFlash suite.
 - Added one DSpark end-to-end test covering a mixed greedy/sampling batch.
 - Ran focused formatting, lint, compilation, and diff checks.

## Performance considerations

Capture remains opt-in. Support is bounded by finite top-k and packed before device-to-host transfer. DSpark performs opt-in target-policy recomputation after verification, so mask capture adds additional probability reconstruction work; this does not affect requests that do not request masks.

## Accuracy Tests

Targeted final implementation:

- 4/4 DFlash sampling-mask unit tests passed in Modal:
    - tensor-backed deferred materialization;
    - filtered target probabilities and selected-token logprobs;
    - greedy singleton masks;
    - whole-request rejection on support overflow.

- Added registered DFlash server coverage asserting:
    - seven generated tokens produce seven masks and seven logprobs;
    - every sampled token belongs to its corresponding support.

- Full Qwen3.6 NVFP4+DFlash B200 run:
    - 98,304 measured output tokens;
    - 49,152 requested masks returned, with no missing masks;
    - maximum observed top-p support was 5,306; a repeat reached 7,476;
    - completed without generation or mask-alignment errors.

This feature only returns metadata and is not intended to change sampling decisions, so no downstream task-accuracy benchmark was run.

## Speed Tests and Profiling

One B200 with:

- nvidia/Qwen3.6-35B-A3B-NVFP4;
- modal-labs/Qwen3.6-35B-A3B-DFlash;
- DFlash block size 8, FA4 draft attention;
- 32 concurrent requests, 512 generated tokens each;
- warmup followed by six alternating mask/no-mask rounds.

Results:

- Without masks: 4,163.78 tok/s median
- With masks: 4,160.45 tok/s median
- Ratio: 0.9992, or approximately −0.08%

The difference was within run-to-run noise. A complete repeat also showed no slowdown.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35150642910](https://github.com/sgl-project/sglang/actions/runs/35150642910)
Latest PR Test (Extra): :x: [Run #35150642360](https://github.com/sgl-project/sglang/actions/runs/35150642360)
Latest PR Test (AMD ROCm 10): :x: [Run #35150642713](https://github.com/sgl-project/sglang/actions/runs/35150642713)

### CODE REVIEW

**Comment by dcw02** [comment]
/rerun-test test/registered/sampling/test_sampling_mask.py test/registered/spec/dflash/test_dflash.py test/registered/core/test_basic_sanity_dspark.py

**Comment by dcw02** [comment]
/rerun-test test/registered/sampling/test_sampling_mask.py test/registered/spec/dflash/test_dflash.py test/registered/core/test_basic_sanity_dspark.py

**Comment by gilfordting** [comment]
/rerun-test test/registered/sampling/test_sampling_mask.py test/registered/spec/dflash/test_dflash.py test/registered/core/test_basic_sanity_dspark.py

**Comment by gilfordting** [comment]
/rerun-test test/registered/sampling/test_sampling_mask.py test/registered/spec/dflash/test_dflash.py test/registered/core/test_basic_sanity_dspark.py

**Comment by gilfordting** [comment]
/rerun-test test/registered/sampling/test_sampling_mask.py test/registered/spec/dflash/test_dflash.py test/registered/core/test_basic_sanity_dspark.py
